File: 589. N-ary Tree Preorder Traversal/solution.js
/**
 * // Definition for a Node.
 * function Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function(root) {
  const result = [];
  const dfs = node => {
    if (node) {
      result.push(node.val);
      node.children.forEach(el => dfs(el));
    }
  };
  dfs(root);
  return result;
};

File: 589. N-ary Tree Preorder Traversal/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        self.ret = []
        self.helper(root)
        return self.ret

    def helper(self, node: 'Node'):
        if not node:
            return
        self.ret.append(node.val)
        for n in node.children:
            self.helper(n)
